mir-ci/mir_ci/benchmarker.py

```diff
@@ -0,0 +1,134 @@
+
+    async def _run(self) -> None:
+        while self.running:
+            try:
+                self.backend.poll()
+            except:
+                pass
+            await asyncio.sleep(self.poll_time_seconds)
+
+    async def start(self) -> None:
+        if self.running is True:
+            return
+
+        self.running = True
+        for program in self.programs:
+            await program.__aenter__()
+            self.backend.add(program)
+
+        self.task = asyncio.ensure_future(self._run())
+
+    async def stop(self) -> None:
+        if self.running is False:
+            return
+        
+        self.running = False
+        if self.task:
+            self.task.cancel()
+            with suppress(asyncio.CancelledError):
+                await self.task
+
+        for program in self.programs:
+            await program.__aexit__()
+    
+    async def __aenter__(self):
+        await self.start()
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    def generate_report(self, record_property: Callable[[str, object], None]) -> None:
+        self.backend.generate_report(record_property)
+
+
+class CgroupsBackend(BenchmarkBackend):
+    class ProcessInfo:
+        name: str
+        start_time_seconds: float
+        cpu_time_microseconds_total: float
+        mem_bytes_total: int
+        mem_bytes_max: int
+        num_data_points: int
+
+        def __init__(self, name: str) -> None:
+            self.name = name
+            self.start_time_seconds = time.time()
+            self.cpu_time_microseconds_total = 0
+            self.mem_bytes_total = 0
+            self.mem_bytes_max = 0
+            self.num_data_points = 0
+        
+    def __init__(self) -> None:
+        self.data_records: Dict[Benchmarkable, CgroupsBackend.ProcessInfo] = {}
+
+    def add(self, program: Benchmarkable) -> None:
+        self.data_records[program] = CgroupsBackend.ProcessInfo(program.get_name())
+    
+    def poll(self) -> None:
+        for program in self.data_records:
+            try:
+                cgroup = program.get_cgroup()
+                current_memory_bytes = cgroup.get_current_memory()
+                cpu_time_microseconds_total = cgroup.get_cpu_time_microseconds()
+                peak_memory_bytes = cgroup.get_peak_memory()
+                self.data_records[program].cpu_time_microseconds_total = cpu_time_microseconds_total
+                self.data_records[program].mem_bytes_total += current_memory_bytes
+
+                if peak_memory_bytes is not None:
+                    self.data_records[program].mem_bytes_max = peak_memory_bytes
+                elif self.data_records[program].mem_bytes_max < current_memory_bytes:
+                    self.data_records[program].mem_bytes_max = current_memory_bytes
+            
+                self.data_records[program].num_data_points += 1
+            except (RuntimeError, FileNotFoundError, PermissionError, OSError) as ex:
```

Comment:
Do we really want to ignore all these? If we're waiting for the correct cgroup in `get_cgroup()` now, then none of these would be expected when things are going well?

Reply:
This is still pending.

Reply:
Sounds good!